engine/src/host_imports/utils.rs

```diff
@@ -0,0 +1,165 @@
+    {
+        return Err(ResolveError::Signature {
+            expected: (*export_func_signature).clone(),
+            found: Args::types().to_vec(),
+        });
+    }
+
+    let func_wasm_inner = module_inner
+        .runnable_module
+        .get_trampoline(&module_inner.info, export_func_signature_idx)
+        .unwrap();
+
+    let export_func_ptr = match export_func_index.local_or_import(&module_inner.info) {
+        LocalOrImport::Local(local_func_index) => module_inner
+            .runnable_module
+            .get_func(&module_inner.info, local_func_index)
+            .unwrap(),
+        _ => {
+            return Err(ResolveError::ExportNotFound {
+                name: name.to_string(),
+            })
+        }
+    };
+
+    let typed_func: Func<'_, Args, Rets, wasmer_core::typed_func::Wasm> =
+        Func::from_raw_parts(func_wasm_inner, export_func_ptr, None, ctx as _);
+
+    Ok(typed_func)
+}
+
+pub(super) fn write_to_wasm_mem(context: &mut Ctx, address: usize, value: &[u8]) {
+    let memory = context.memory(0);
+
+    memory.view::<u8>()[address..(address + value.len())]
+        .iter()
+        .zip(value.iter())
+        .for_each(|(cell, byte)| cell.set(*byte));
+}
+
+pub(super) fn itypes_args_to_wtypes(itypes: &[IType]) -> Vec<WType> {
+    itypes
+        .iter()
+        .map(|itype| match itype {
+            IType::F32 => vec![WType::F32],
+            IType::F64 => vec![WType::F64],
+            IType::I64 | IType::U64 => vec![WType::I64],
+            IType::String | IType::Array(_) => vec![WType::I32, WType::I32],
+            _ => vec![WType::I32],
+        })
+        .flatten()
+        .collect::<Vec<_>>()
+}
+
+pub(super) fn itypes_output_to_wtypes(itypes: &[IType]) -> Vec<WType> {
+    itypes
+        .iter()
+        .map(|itype| match itype {
+            IType::F32 => vec![WType::F32],
+            IType::F64 => vec![WType::F64],
+            IType::I64 | IType::U64 => vec![WType::I64],
+            IType::String | IType::Array(_) | IType::Record(_) => vec![],
+            _ => vec![WType::I32],
+        })
+        .flatten()
+        .collect::<Vec<_>>()
+}
+
+#[macro_export] // https://github.com/rust-lang/rust/issues/57966#issuecomment-461077932
+/// Initialize Wasm function in form of Box<RefCell<Option<Func<'static, args, rets>>>> only once.
+macro_rules! init_wasm_func_once {
+    ($func:ident, $ctx:ident, $args:ty, $rets:ty, $func_name:ident, $ret_error_code: expr) => {
+        if $func.borrow().is_none() {
+            let raw_func = match unsafe {
+                super::utils::get_export_func_by_name::<$args, $rets>($ctx, $func_name)
+            } {
+                Ok(func) => func,
+                Err(_) => return vec![WValue::I32($ret_error_code)],
+            };
+
+            unsafe {
+                // assumed that this function will be used only in the context of closure
+                // linked to a corresponding Wasm import, so it is safe to make is static
```

Comment:
Maybe elaborate a little here? Like, what's about the Wasm import, is it considered to be static (as in live until the end of the program)? If so, maybe point out where is it stored? Like:
```suggestion
                // assumed that this function will be used only in the context of closure
                // linked to a corresponding Wasm import, so it is safe to make is static
                // because all Wasm imports live in the Wasmer instances, which 
                // is itself static (i.e., lives until the end of the program)
```